packages/plugins/persisted-operations/src/jsonFilesStore.ts

```diff
@@ -0,0 +1,63 @@
+/* eslint-disable no-console */
+import { basename, extname, isAbsolute, resolve } from 'path';
+import { readFile } from 'fs/promises';
+import { PersistedOperationsStore } from './plugin';
+
+export class JsonFilesStore {
```

Comment:
To me, this class feels way more complicated and confusing than it must be. A Store on which you call a get method to get a Store?

I don't think we need to provide a primitive for loading multiple JSON files and should be up to the user to implement.

Something basic like the following could be more straightforward.

```ts
class JSONStore extends Map {
  async load(path: string) {
    // logic for loading path
   const data = loadJSONFile(path)
   for (const [key, value] of Object.entries(operationsObject)) {
      this.set(key, value);
    }
  }
}
```

With the following usage:

```ts
const persistedOperationsPaths = [
  resolve(process.cwd(), 'assets/client1PersistedOperations.json'),
  resolve(process.cwd(), 'assets/client2PersistedOperations.json'),
];

const store = new JSONStore();
for (const path of persistedOperationsPaths) {
  store.load(path);
}
 // load and parse persisted-operations files

const getEnveloped = envelop({
  plugins: [
    // ... other plugins ...
    usePersistedOperations({
      store,
    }),
  ],
});
```

And if you need to use a different store, based on context:

```ts
const persistedOperationsPaths = [
  resolve(process.cwd(), 'assets/client1PersistedOperations.json'),
  resolve(process.cwd(), 'assets/client2PersistedOperations.json'),
];

const store1 = new JSONStore();
const store2 = new JSONStore();

store1.load(path[0]);
store2.load(path[1]);
 // load and parse persisted-operations files

const getEnveloped = envelop({
  plugins: [
    // ... other plugins ...
    usePersistedOperations({
      store: ctx => ctx.sth ? store1 : store2,
    }),
  ],
});
```

Reply:
> this class feels way more complicated and confusing than it must be

Possible, but my goal with this class is not just to be something that people can use, but also to give examples of how you can build full solutions with more complex scenarios, such as handling multiple lists with a clue on how you can update them (all or some) on the fly.
Even though, yeah, maybe you can achieve that with a much simpler class, at the cost of actually having to initiate multiple instances to serve multiple stores.

Maybe we can commit a more minimal setup and this more complex solution could live in a guide or a recipe.

> A Store on which you call a get method to get a Store?

To be more precise, a store which you call a get method to get the Map of persisted operations that are relevant to the requested operation, which might be only from a single list of those supported by your store.

> should be up to the user to implement.

Yes, but don’t forget that you don’t have to use this class in order to use the plugin.
Users are still free to implement what works for them.
This works, it handles a single list scenario, more complex scenarios with multiple lists and live updates; but it's also there as a reference.

> Something basic like the following could be more straightforward

It would.
But personally, I would not extend from Map.
At that stage you return a Map that people can mutate directly by invoking all the methods available, including `set`; the risk is much bigger as you might end up with something that goes well beyond storing entries from a designated static file.
This is also the reason why in my implementation I have a method `get` that returns the Map, and I keep the store as private; I want to enforce immutability patterns and make it straightforward that you retrieve the Map and so this is meant to be read-only.

Anyway, I am not precious about the jsonFilesStore.
If the feedback is to have something that is much simpler and we fully leave it to the users to implement more complex scenarios on their own, I will be ok with that.
The more complex store implementation might well live in a guide or blog post, just to inspire or provide some guidance.